File: daily_problems/2026/06/0608/personal_submission/cf102441f_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

long long quickPow(long long base, long long power, long long mod) {
	if (power == 0) return 1 % mod;
	long long cur = quickPow(base, power / 2, mod);
	return power & 1 ? base * cur % mod * cur % mod : cur * cur % mod; 
}

int cnt[30][30][2][2];

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, x, y;
	cin >> n >> x >> y;

	for (int i = 0; i < n; i ++) {
		int v;
		cin >> v;

		for (int j = 0; j < 30; j ++) {
			for (int k = 0; k < 30; k ++) {
				cnt[j][k][v >> j & 1][v >> k & 1] ++;
			}
		}
	}

	int mod = 1e9 + 7;

	int p1 = x * quickPow(y, mod - 2, mod), p0 = (mod + 1 - p1) % mod;

	vector<int> probs(n + 1, 0);
	int v1 = 0, v0 = 1;

	for (int i = 1; i <= n; i ++) {
		int nv1 = (1ll * v1 * p0 + 1ll * v0 * p1) % mod;
		int nv0 = (1ll * v0 * p0 + 1ll * v1 * p1) % mod;

		v1 = nv1;
		v0 = nv0;

		probs[i] = v1;
	}

	int ans = 0;

	for (int i = 0; i < 30; i ++) {
		for (int j = 0; j < 30; j ++) {
			vector<int> dp(4, 0);
			dp[0] = 1;

			for (int vi = 0; vi < 2; vi ++) {
				for (int vj = 0; vj < 2; vj ++) {
					int v = vi * 2 + vj;
					int x = cnt[i][j][vi][vj];

					int p1 = probs[x];
					int p0 = (mod + 1 - p1) % mod;

					vector<int> ndp(4, 0);

					for (int idx = 0; idx < 4; idx ++) {
						ndp[idx] = (1ll * dp[idx ^ v] * p1 + 1ll * dp[idx] * p0) % mod;
					}

					dp.swap(ndp);
				}
			}
			ans += (1ll << i + j) % mod * dp[3] % mod;
			ans %= mod;
		}
	}

	cout << ans;

	return 0;
}

File: daily_problems/2026/06/0608/personal_submission/cf102441f_yawn_sean.py
# Submission link: https://codeforces.com/gym/102441/submission/377639680
def main():
    mod = 10 ** 9 + 7
    
    n, x, y = MII()
    nums = LII()
    
    v1 = x * pow(y, -1, mod) % mod
    v0 = (mod + 1 - v1) % mod
    
    cnt = [[0] * 900 for _ in range(4)]
    
    for v in nums:
        vals = [v >> i & 1 for i in range(30)]
        
        for i in range(30):
            for j in range(30):
                cnt[vals[i] * 2 + vals[j]][i * 30 + j] += 1
    
    p0, p1 = 1, 0
    saved = [0] * (n + 1)
    
    for i in range(1, n + 1):
        p0, p1 = (p0 * v0 + p1 * v1) % mod, (p1 * v0 + p0 * v1) % mod
        saved[i] = p1
    
    ans = 0
    
    for i in range(30):
        for j in range(30):
            dp = [0] * 4
            dp[0] = 1
            
            for vi in range(2):
                for vj in range(2):
                    v = vi * 2 + vj
                    x = cnt[v][i * 30 + j]
                    
                    p1 = saved[x]
                    p0 = (mod + 1 - p1) % mod
                    
                    ndp = [0] * 4
                    
                    for idx in range(4):
                        ndp[idx] = (dp[idx ^ v] * p1 + dp[idx] * p0) % mod
                    
                    dp = ndp
            
            ans += (1 << i + j) % mod * dp[3] % mod
            ans %= mod
    
    print(ans)